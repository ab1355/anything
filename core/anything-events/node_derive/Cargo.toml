[package]
name = "node_derive"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html


[lib]
proc-macro = true
path = "src/lib.rs"

[dependencies]
quote = "1.0"
syn = { version = "2.0", features = [
    "extra-traits",
    "full",
    "derive",
    "parsing",
] }
proc-macro2 = "1.0"
proc-macro-error = "1"


[dev-dependencies]
crossbeam = "0.8"
crossbeam-channel = "0.5"
trybuild = "1"
