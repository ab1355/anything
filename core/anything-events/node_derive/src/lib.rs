#![recursion_limit = "128"]
extern crate proc_macro;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

enum FieldType {
    Input,
    Output,
    State,
}

struct ParsedFields<'a> {
    recv_fields: Vec<&'a syn::Field>,
    send_fields: Vec<&'a syn::Field>,
}

#[proc_macro_derive(Node, attributes(aggregate, pass_by_ref))]
pub fn node_derive(stream: TokenStream) -> TokenStream {
    let ast = parse_macro_input!(stream as DeriveInput);
    let name = &ast.ident;
    let attributes = &ast.attrs;
    let generics = &ast.generics;
    let data = &ast.data;
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    let mut aggregate = false;
    let mut pass_by_ref = false;

    for attr in attributes {
        let _ = attr.parse_nested_meta(|meta| {
            if meta.path.is_ident("aggregate") {
                aggregate = true;
                return Ok(());
            }

            if meta.path.is_ident("pass_by_ref") {
                pass_by_ref = true;
                return Ok(());
            }

            return Err(meta.error("unrecognized attribute"));
        });
    }

    let recv_fields;
    let send_fields;

    match data {
        syn::Data::Struct(data_struct) => match &data_struct.fields {
            syn::Fields::Named(fields) => {
                let parsed_fields = parse_fields(fields);
                recv_fields = parsed_fields.recv_fields.clone();
                send_fields = parsed_fields.send_fields.clone();
            }
            _ => {
                let err = quote! {
                    compile_error!("can't derive node for non-named structs");
                };
                return proc_macro::TokenStream::from(err);
            }
        },
        _ => {
            let err = quote! {
                compile_error!("can't derive node for non-named structs");
            };
            return proc_macro::TokenStream::from(err);
        }
    }

    if recv_fields.is_empty() && send_fields.is_empty() {
        let err = quote! {
        compile_error!("node needs at least one NodeReceiver or \
            NodeSender");
        };
        return proc_macro::TokenStream::from(err);
    }

    let recv_idents: Vec<syn::Ident> = recv_fields
        .iter()
        .map(|x| x.ident.clone().unwrap())
        .collect();

    let send_idents: Vec<syn::Ident> = send_fields
        .iter()
        .map(|x| x.ident.clone().unwrap())
        .collect();

    let send_idents1 = &send_idents;
    let send_idents2 = &send_idents;
    let recv_block_idents = &recv_idents;
    let recv_block_fields = &recv_idents;

    let start = quote! {
        fn start(&mut self) {
            #(
                for (send, val) in &self.#send_idents2 {
                    match val {
                        Some(v) => send.send(v.clone()).unwrap(),
                        None => continue,
                    }
                }
            )*
            loop {
                if self.call().is_err() {
                    break;
                }
            }
        }
    };

    let run_func = if pass_by_ref {
        quote! {
            let res = self.run(#(&#recv_block_idents),*)?;
        }
    } else {
        quote! {
            let res = self.run(#(#recv_block_idents),*)?;
        }
    };

    let send_func = if aggregate {
        quote! {
            if let Some(res) = res {
                #(
                    for (send, _) in &self.#send_idents1 {
                        match send.send(res.clone()) {
                            Ok(_) => (),
                            Err(e) => return Err(NodeError::CommError),
                        }
                    }
                )*
            }
        }
    } else {
        quote! {
            #(
                for (send, _) in &self.#send_idents1 {
                    match send.send(res.clone()) {
                        Ok(_) => (),
                        Err(e) => return Err(NodeError::CommError),
                    }
                }
            )*
        }
    };

    let call = quote! {
        fn call(&mut self) -> Result<(), NodeError> {
            #(
                let #recv_block_idents = match self.#recv_block_fields {
                    Some(ref r) => r.recv().or(Err(NodeError::DataEnd))?,
                    None => return Err(NodeError::PermanentError),
                };
            )*
            #run_func
            #send_func
            Ok(())
        }
    };

    let is_connected = quote! {
        fn is_connected(&self) -> bool {
            #(
                if self.#recv_block_fields.is_none() {
                    return false;
                }
            )*
            #(
                if self.#send_idents1.is_empty() {
                    return false;
                }
            )*
            return true;
        }
    };

    let expanded = quote! {
        impl #impl_generics Node for #name #ty_generics #where_clause {
            #start
            #call
            #is_connected
        }
    };

    TokenStream::from(expanded)
}

fn parse_fields(fields: &syn::FieldsNamed) -> ParsedFields {
    let mut recv_fields = vec![];
    let mut send_fields = vec![];
    for field in &fields.named {
        match parse_type(&field) {
            FieldType::Input => recv_fields.push(field),
            FieldType::Output => send_fields.push(field),
            _ => continue,
        }
    }
    ParsedFields {
        recv_fields,
        send_fields,
    }
}

fn parse_type(field: &syn::Field) -> FieldType {
    let ty = &field.ty;
    let type_str = quote! {#ty}.to_string();
    if type_str.starts_with("NodeReceiver") {
        FieldType::Input
    } else if type_str.starts_with("NodeSender") {
        FieldType::Output
    } else {
        FieldType::State
    }
}
