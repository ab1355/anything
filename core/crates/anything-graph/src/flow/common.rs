use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Builder)]
#[builder(setter(into, strip_option), default)]
pub struct PackageData {
    pub id: Uuid,
    pub label: String,
    pub author: Option<String>,
    pub version: Option<String>,
}

impl Default for PackageData {
    fn default() -> Self {
        Self {
            id: Uuid::new_v4(),
            label: String::default(),
            author: None,
            version: None,
        }
    }
}
