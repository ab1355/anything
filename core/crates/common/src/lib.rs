pub mod config;
pub mod de;
pub mod error;
pub mod hashing;
pub mod posix;
pub mod spawning;
pub mod tracing;
pub mod utils;

pub use config::*;
pub use hashing::*;
pub use posix::*;
pub use spawning::*;
pub use crate::tracing::setup_tracing;
pub use utils::*;
