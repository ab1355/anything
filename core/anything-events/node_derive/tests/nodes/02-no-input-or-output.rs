mod common;
use common::*;
use node_derive::Node;

#[derive(Node)]
pub struct SimpleNode {}

impl SimpleNode {
    pub fn new() -> Self {
        Self {}
    }

    pub fn run(&mut self, _x: u32) -> Result<Vec<u32>, NodeError> {
        Ok(vec![1])
    }
}

fn main() {}
