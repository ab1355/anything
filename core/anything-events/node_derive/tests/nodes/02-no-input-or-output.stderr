error: node needs at least one NodeReceiver or NodeSender
 --> tests/nodes/02-no-input-or-output.rs:5:10
  |
5 | #[derive(Node)]
  |          ^^^^
  |
  = note: this error originates in the derive macro `Node` (in Nightly builds, run with -Z macro-backtrace for more info)
